service events - error during locator initialization

#### Description 
The error occurs during the startup phase of the REST API. The log reports the following:

> Caused by: com.google.inject.ProvisionException: Unable to provision, see the following errors:
>  
> 1) Error injecting constructor, java.lang.NullPointerException
>   at org.eclipse.kapua.service.user.internal.UserServiceImpl.<init>(UserServiceImpl.java:72)
>   while locating org.eclipse.kapua.service.user.internal.UserServiceImpl
>   at org.eclipse.kapua.locator.guice.KapuaModule.configure(KapuaModule.java:121)
>   while locating org.eclipse.kapua.service.user.UserService
>     for field at org.eclipse.kapua.service.user.internal.UserServiceModule.userService(UserServiceModule.java:27)
>   while locating org.eclipse.kapua.service.user.internal.UserServiceModule
>   while locating org.eclipse.kapua.commons.core.ServiceModule annotated with @com.

The error is due  to the fact that the list of service modules are retrieved (and instantiated) before completing the initialization of the locator (for the Guice implementation). The retrieve of the service modules triggers the initialization of some service implementations which in turn try to access the locator.

#### Expected behavior
Objects managed by the Kapua locator do not have to be created before the locator itself is completely initialized (for the Guice implementation). The solution is to defer the instantiation of any object until the locator is fully initialized.

--- a/commons/src/main/java/org/eclipse/kapua/commons/core/ServiceModuleConfiguration.java
+++ b/commons/src/main/java/org/eclipse/kapua/commons/core/ServiceModuleConfiguration.java
@@ -16,16 +16,20 @@ import java.util.Set;
 
 public class ServiceModuleConfiguration {
 
-    private static ServiceModuleProvider serviceModuleProvider;
+    public interface ConfigurationProvider {
+        ServiceModuleProvider get() ;
+    }
+
+    private static ConfigurationProvider cofigurationProvider;
 
     private ServiceModuleConfiguration() {}
 
-    public static void setModuleProvider(ServiceModuleProvider aServiceModuleProvider) {
-        serviceModuleProvider = aServiceModuleProvider;
+    public static void setConfigurationProvider(ConfigurationProvider aProvider) {
+        cofigurationProvider = aProvider;
     }
 
     public static Set<ServiceModule> getServiceModules() {
-        return serviceModuleProvider.getModules();
+        return cofigurationProvider.get().getModules();
     }
 
  }

--- a/locator/guice/src/main/java/org/eclipse/kapua/locator/guice/GuiceLocatorImpl.java
+++ b/locator/guice/src/main/java/org/eclipse/kapua/locator/guice/GuiceLocatorImpl.java
@@ -39,12 +39,16 @@ public class GuiceLocatorImpl extends KapuaLocator {
 
     public GuiceLocatorImpl() {
         injector = Guice.createInjector(new KapuaModule());
-        ServiceModuleConfiguration.setModuleProvider(injector.getInstance(ServiceModuleProvider.class));
+        ServiceModuleConfiguration.setConfigurationProvider(() -> {
+            return injector.getInstance(ServiceModuleProvider.class);
+        });
     }
 
     public GuiceLocatorImpl(String resourceName) {
         injector = Guice.createInjector(new KapuaModule(resourceName));
-        ServiceModuleConfiguration.setModuleProvider(injector.getInstance(ServiceModuleProvider.class));
+        ServiceModuleConfiguration.setConfigurationProvider(() -> {
+            return injector.getInstance(ServiceModuleProvider.class);
+        });
     }
 
     @Override

